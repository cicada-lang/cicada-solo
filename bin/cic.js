#!/usr/bin/env node

const process = require("process")

process.on("unhandledRejection", (error) => {
  console.error(error)
  process.exit(1)
})

const { createCommandRunner } = require("../lib/console")

createCommandRunner().run()
