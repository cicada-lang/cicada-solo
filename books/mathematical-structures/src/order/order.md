---
title: Order Theory
---

# Introduction

> Order theory is a branch of mathematics which investigates the
> intuitive notion of order using binary relations. It provides a formal
> framework for describing statements such as "this is less than that"
> or "this precedes that". This article introduces the field and
> provides basic definitions. A list of order-theoretic terms can be
> found in the order theory glossary.
>
> -- [wikipedia / Order theory](https://en.wikipedia.org/wiki/Order_theory)


# PreOrder

We Start our definitions from `PreOrder`.

``` cicada
class PreOrder {
  Element: Type
  Under(Element, Element): Type

  reflexive(x: Element): Under(x, x)
  transitive(
    x: Element,
    y: Element,
    z: Element,
    Under(x, y),
    Under(y, z),
  ): Under(x, z)
}
```

A preorder is a thin category -- `hom_set(A, B).size <= 1`.

# PartialOrder

From `PreOrder`, we can add antisymmetric to get `PartialOrder`:

``` cicada
class PartialOrder extends PreOrder {
  antisymmetric(
    x: Element,
    y: Element,
    Under(x, y),
    Under(y, x),
  ): Equal(Element, x, y)
}
```

Note that, there will be no cycle in a partial order,
and we can implement topological sort for a partial order.

# Equivalence

Again from `PreOrder`, we can get equivalence relation by adding symmetric.

``` cicada

class Equivalence extends PreOrder {
  symmetric(x: Element, y: Element, Under(x, y)): Under(y, x)

  // TODO Maybe we should rename `Under` to `Eq`
  // Eq: Type = Under
}
```
