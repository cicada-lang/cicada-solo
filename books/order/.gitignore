# npm
node_modules

# emacs
*~
*#*
.#*
