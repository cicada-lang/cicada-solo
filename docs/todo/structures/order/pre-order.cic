@module structure.order

// NOTE Thin category.
// NOTE hom_set(A, B).size <= 1
@class PreOrder {
  Elem: Type
  Under(Elem, Elem): Type

  reflexive(x: Elem): Under(x, x)

  @given x: Elem, y: Elem, z: Elem
  transitive(Under(x, y), Under(y, z)): Under(x, z)
}

@class StrictUnder {
  order: PreOrder
  x: order.Elem
  y: order.Elem
  strict: NonEqual(x, y)
  under: order.Under(x, y)
}
