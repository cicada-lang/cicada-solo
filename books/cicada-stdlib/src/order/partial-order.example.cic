import { PartialOrder } from "./partial-order.cic"

PartialOrder
