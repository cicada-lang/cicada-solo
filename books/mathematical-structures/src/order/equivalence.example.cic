import { Equivalence } from "./equivalence.cic"

Equivalence
