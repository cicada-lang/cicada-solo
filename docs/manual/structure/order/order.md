---
title: Order Theory
author: Xie Yuheng
date: 2021-10-20
---

# Introduction

> Order theory is a branch of mathematics which investigates the
> intuitive notion of order using binary relations. It provides a formal
> framework for describing statements such as "this is less than that"
> or "this precedes that".
>
> -- [Wikipedia / Order theory](https://en.wikipedia.org/wiki/Order_theory)

# PreOrder

Order theory focus on transitive binary relations,
we start our definitions from a basic one -- `PreOrder`.

```cicada
import { Set } from "../set/set.md"
import { Not } from "../logic/basic.cic"

class PreOrder {
  Element: Type

  element_set: Set(Element)

  Under(Element, Element): Type

  reflexive(x: Element): Under(x, x)

  transitive(
    implicit x: Element,
    implicit y: Element,
    Under(x, y),
    implicit z: Element,
    Under(y, z),
  ): Under(x, z)

  NotComparable(x: Element, y: Element): Type {
    return Both(Not(Under(x, y)), Not(Under(y, x)))
  }
}
```

A preorder is a thin category -- `hom_set(A, B).size <= 1`.

# Order

From `PreOrder`, we can add antisymmetric to get `Order`:

```cicada
class Order extends PreOrder {
  antisymmetric(
    implicit x: Element,
    implicit y: Element,
    Under(x, y),
    Under(y, x),
  ): element_set.Eq(x, y)
}
```

Note that, there will be no cycle in a partial order,
and we can implement topological sort for a partial order.

# TotalOrder

```cicada
import { Either } from "../datatype/either.cic"

class TotalOrder extends Order {
  totality(x: Element, y: Element): Either(Under(x, y), Under(y, x))
}
```

# Equivalence

From `PreOrder` again, we can get equivalence relation by adding symmetric.

```cicada
class Equivalence extends PreOrder {
  symmetric(
    implicit x: Element,
    implicit y: Element,
    Under(x, y),
  ): Under(y, x)

  // TODO Maybe we should rename `Under` to `Eq`
  // Eq: Type = Under
}
```
