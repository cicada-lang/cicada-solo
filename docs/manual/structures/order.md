---
title: Order Theory
---

# Introduction

> Order theory is a branch of mathematics which investigates the
> intuitive notion of order using binary relations. It provides a formal
> framework for describing statements such as "this is less than that"
> or "this precedes that".
>
> -- [Wikipedia / Order theory](https://en.wikipedia.org/wiki/Order_theory)

# PreOrder

Order theory focus on transitive binary relations,
we start our definitions from a basic one -- `PreOrder`.

```cicada
import { Not } from "../logic/basic.cic"

class PreOrder {
  Element: Type

  Under(Element, Element): Type

  reflexive(x: Element): Under(x, x)

  transitive(
    implicit x: Element,
    implicit y: Element,
    Under(x, y),
    implicit z: Element,
    Under(y, z),
  ): Under(x, z)

  NotComparable(x: Element, y: Element): Type {
    return Both(Not(Under(x, y)), Not(Under(y, x)))
  }
}
```

A preorder is a thin category -- `hom_set(A, B).size <= 1`.

# Order

From `PreOrder`, we can add antisymmetric to get `Order`:

```cicada
class Order extends PreOrder {
  antisymmetric(
    implicit x: Element,
    implicit y: Element,
    Under(x, y),
    Under(y, x),
  ): Equal(Element, x, y)
}
```

Note that, there will be no cycle in a partial order,
and we can implement topological sort for a partial order.

# TotalOrder

```cicada
import { Either } from "../datatypes/either.md"

class TotalOrder extends Order {
  totality(x: Element, y: Element): Either(Under(x, y), Under(y, x))
}
```

# Equivalence

From `PreOrder` again, we can get equivalence relation by adding symmetric.

```cicada
class Equivalence extends PreOrder {
  symmetric(
    implicit x: Element,
    implicit y: Element,
    Under(x, y),
  ): Under(y, x)

  // TODO Maybe we should rename `Under` to `Eq`
  // Eq: Type = Under
}
```
