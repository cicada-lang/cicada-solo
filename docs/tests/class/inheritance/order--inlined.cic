class OrderInlined {
  Element: Type
  Under(Element, Element): Type

  reflexive(x: Element): Under(x, x)

  transitive(
    implicit x: Element,
    implicit y: Element,
    Under(x, y),
    implicit z: Element,
    Under(y, z),
  ): Under(x, z)

  antisymmetric(
    implicit x: Element,
    implicit y: Element,
    Under(x, y),
    Under(y, x),
  ): Equal(Element, x, y)
}

// NOTE Inlined definition is the equal to definition by inheritance.

import { Order } from "./order.cic"

compute the(Equal(Type, Order, OrderInlined), refl)
