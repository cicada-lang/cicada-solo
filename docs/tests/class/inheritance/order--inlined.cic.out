refl(vague Type, vague class {
  Element: Type
  Under: (Element, Element) -> Type
  reflexive: (x: Element) -> Under(x, x)
  transitive: (implicit x: Element, implicit y: Element, Under(x, y), implicit z: Element, Under(y, z)) -> Under(x, z)
  antisymmetric: (implicit x: Element, implicit y: Element, Under(x, y), Under(y, x)) -> Equal(Element, x, y)
}): Equal(Type, class {
  Element: Type
  Under: (Element, Element) -> Type
  reflexive: (x: Element) -> Under(x, x)
  transitive: (implicit x: Element, implicit y: Element, Under(x, y), implicit z: Element, Under(y, z)) -> Under(x, z)
  antisymmetric: (implicit x: Element, implicit y: Element, Under(x, y), Under(y, x)) -> Equal(Element, x, y)
}, class {
  Element: Type
  Under: (Element, Element) -> Type
  reflexive: (x: Element) -> Under(x, x)
  transitive: (implicit x: Element, implicit y: Element, Under(x, y), implicit z: Element, Under(y, z)) -> Under(x, z)
  antisymmetric: (implicit x: Element, implicit y: Element, Under(x, y), Under(y, x)) -> Equal(Element, x, y)
})
